### PR TITLE
stdlib: faster kronecker product between hermitian and symmetric matrices

### PR DESCRIPTION
**Pull request by araujoms**
The kronecker product between complex hermitian matrices is again hermitian, so it can be computed much faster by only doing the upper (or lower) triangular. As @andreasnoack will surely notice, this only true for types where `conj(a*b) == conj(a)*conj(b)`, so I'm restricting the function to act only on real and complex numbers. In the symmetric case, however, no additional assumption is needed, so I'm letting it act on anything.

Benchmarking showed that the code is roughly 2 times as fast as the vanilla kronecker product, as expected. The fastest case was always the UU case, and the slowest the LU case. The code I used is below
```julia
using LinearAlgebra
using BenchmarkTools
using Quaternions

randrmatrix(d, uplo = :U) = Hermitian(randn(Float64, d, d), uplo)
randcmatrix(d, uplo = :U) = Hermitian(randn(ComplexF64, d, d), uplo)
randsmatrix(d, uplo = :U) = Symmetric(randn(ComplexF64, d, d), uplo)
randqmatrix(d, uplo = :U) = Symmetric(randn(QuaternionF64, d, d), uplo)

dima = 69
dimb = 71
for randmatrix in [randrmatrix, randcmatrix, randsmatrix, randqmatrix]
    for auplo in [:U, :L]
        for buplo in [:U, :L]
            a = randmatrix(dima, auplo)
            b = randmatrix(dimb, buplo)
            c = kron(a,b)
            therm = @belapsed kron!($c, $a, $b)
            C = Matrix(c)
            A = Matrix(a)
            B = Matrix(b)
            told = @belapsed kron!($C, $A, $B)
            @show told/therm
        end
    end
end
```
Weirdly enough, I got this expected speedup in one of my machines, but when running the benchmark in another I got roughly the same time. I guess that's a bug with `BechmarkTools`, because that's not consistent with the times I get running the functions individually, out of the loop.

Another issue is that although I added a couple of tests, I couldn't get them to run. Perhaps someone here can tell me what's going on? I could run the tests from LinearAlgebra, it's just that editing the files made no difference to what was being run. I did get hundreds of errors from `triangular.jl`, but that's untouched by my code.

### CODE REVIEW

**Comment by barucden** [comment]
> I could run the tests from LinearAlgebra, it's just that editing the files made no difference to what was being run.

Did you change the UUID in `Project.toml` (https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md#contributing-to-the-standard-library)?

**Comment by araujoms** [comment]
Yes, I did. My code runs, I've confirmed it by doing `@edit kron(a,b)`, only the tests get mysteriously ignored.

**Comment by jishnub** [comment]
Are you running the tests from `LinearAlgebra.jl`, based on you changing the UUID? That repo isn't being updated at present. You should build Julia with your changes, and use `include("stdlib/LinearAlgebra/test/symmetric.jl")` to run the tests. Usually building Julia isn't necessary as well if you're using Revise, as you may use `Revise.track(LinearAlgebra)` to include your local changes, but I've encountered some issues with Revise on Julia nightly, so a fresh build might be better.

File: stdlib/LinearAlgebra/src/symmetric.jl
Comment (jishnub) [inline stdlib/LinearAlgebra/src/symmetric.jl:511]: The method should probably be restricted to `Hermitian{<:Union{Real,Complex}, <StridedMatrix}` and similarly for `Similar`, and a fallback implementation of the form `kron!(C.data, A.data, B.data)` may be added. This will ensure that sparse arrays (or other special array types) remain performant, if they provide a `kron!` implementation.

**Comment by araujoms** [comment]
Well yes, I was following the instructions for contributing to a standard library, not to base Julia. In any case it worked now, I've changed back the UUID, built Julia again, now the tests run (and pass).

File: stdlib/LinearAlgebra/src/symmetric.jl
Comment (araujoms) [reply stdlib/LinearAlgebra/src/symmetric.jl:511]: Please forgive me my ignorance as I'm rather new to Julia. You are saying I should change the function declarations to
```julia
function kron(A::Hermitian{T, <:StridedMatrix}, B::Hermitian{S, <:StridedMatrix}) where {T<:Union{Real,Complex},S<:Union{Real,Complex}}
function kron(A::Symmetric{T, <:StridedMatrix}, B::Symmetric{S, <:StridedMatrix}) where {T,S}
```
in order to act only on dense matrices, not sparse ones. Fine, I get this. But then I should also give a fallback implementation, that only acts if the sparse type doesn't provide their own `kron!`? How? I thought the original function declaration was already doing this, as a more specialized function declaration for a specific type would take precedence.

Comment (araujoms) [reply stdlib/LinearAlgebra/src/symmetric.jl:511]: I see, the issue is that a generic Kronecker product over sparse arrays will still be vastly more efficient for Hermitian matrices than my dense version, so we don't want to capture that. Of course, such a sparse type can also implement its own specialization for Hermitian matrices, but that's a different issue.

The best I could come up with was
```julia
for T in [:Symmetric, :(Hermitian{<:Union{Real,Complex}})]
    @eval begin
        function kron!(C::$T, A::$T, B::$T)
            wrapper = Base.typename($T).wrapper
            return wrapper(kron!(C.data, A.data, B.data))
        end
    end
end
```
for the fallback and
```julia
for (T, conj, real) in [(:(Symmetric{<:Any,<:StridedMatrix}), :identity, :identity), (:(Hermitian{<:Union{Real,Complex}, <:StridedMatrix}), :conj, :real)]
    @eval begin
        function kron!(C::$T, A::$T, B::$T)
```
for the signature of the method itself. It's rather hideous. Is there a better solution?

Also, doesn't exactly the same issue applies to `dot`? Is there any package actually implementing faster versions of `dot` and `kron` for sparse arrays?

Comment (jishnub) [reply stdlib/LinearAlgebra/src/symmetric.jl:511]: I take my comment back, as `kron` forwarded to the data isn't the same:
```julia
julia> kron(H1, H2)
4×4 Matrix{Float64}:
 0.343295  0.436362  0.106336   0.135163
 0.436362  0.717831  0.135163   0.222349
 0.106336  0.135163  0.0616077  0.0783094
 0.135163  0.222349  0.0783094  0.128822

julia> Hermitian(kron(H1.data, H2.data))
4×4 Hermitian{Float64, Matrix{Float64}}:
 0.343295  0.436362  0.106336   0.135163
 0.436362  0.717831  0.219534   0.222349
 0.106336  0.219534  0.0616077  0.0783094
 0.135163  0.222349  0.0783094  0.128822
```
I think the current implementation is fine.

Comment (araujoms) [reply stdlib/LinearAlgebra/src/symmetric.jl:511]: Ah yes, of course, that would work only if `H == H.data`. The way it is is slow but at least correct. If a sparse type wants a fast version they need to implement it themselves.

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/symmetric.jl:511]: What happened to the `StridedMatrix` restriction? I think another option for dispatching on special matrix types is to use an unwrapping mechanism as in matmul and write a "symmetric kronecker kernel" or "triangular kronecker kernel", acting on the `data` fields:

```julia
function kron(A::Symmetric, B::Symmetric)
    resultuplo = A.uplo == 'U' || B.uplo == 'U' ? :U : :L
    C = kronprod_dest(A, B)
    return Symmetric(_symkron!(C, A.data, B.data, identity, identity, C.uplo, A.uplo, B.uplo), resultuplo)
end

function _symkron!(C, A, B, conj::TC, real::TR, Cuplo, Auplo, Buplo) where {TC,TR}
...
end
```

... as a proposal/sketch. What do you think? Packages could then overload `_symkron` for their storage types.

Comment (araujoms) [reply stdlib/LinearAlgebra/src/symmetric.jl:511]: Why would that help? Any package can define `kron!(::Hermitian{<:Union{Real,Complex}}, Hermitian{<:Union{Real,Complex}, MyArray}, ::Hermitian{<:Union{Real,Complex}, MyArray})` and that would take precedence, as @jishnub pointed out.

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/symmetric.jl:511]: The issue with that is that the method is dispatched over a type parameter, which, as we learned the lesson in matmul, is adding to package load time and latency. Moreover, you might be able to reuse that "kernel" in the kronecker product of triangular matrices, for instance. So, there is a multi-parameter challenge here: OTOH you want to dispatch on the "triangularity", OTOH you also want to dispatch on the storage type. In matmul, we do that by first unwrapping the outer layer and directing to specially called methods (not just `mul!`, but `generic_matmatmul!` and alike), which in turn are then overloaded for different storage types (dense/sparse/GPU).

Comment (araujoms) [reply stdlib/LinearAlgebra/src/symmetric.jl:511]: Fair enough, I can do that.

I can't use literally the same kernel for triangular matrices for two reasons. First, because they don't use `uplo`, but instead have two different types, `UpperTriangular` and `LowerTriangular`. I can invent a fake `uplo` for them and make it work, though. The second problem is that when they have mismatched `uplo` the result is not triangular, so I have to add another argument to throw an error (or is it enough to leave the error handling to `kron`, and say that anyone using `_symkron!` is on their own?)

Also, I don't understand what you meant with `conj::TC, real::TR`. Shouldn't `conj` and `real` be functions here? What are these types?

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/symmetric.jl:511]: BTW: These are just suggestions, so we might consider them for another PR.

For the triangular ones, one could have, e.g.,

```julia
function kron(A::UpperTriangular, B::UpperTriangular)
    # create result storage C
    return UpperTriangular(_symkron(C, A.data, B.data, identity, identity, 'U', 'U', 'U'))
end
```

> Also, I don't understand what you meant with conj::TC, real::TR. Shouldn't conj and real be functions here? What are these types?

Yes, they are functions, and I believe we would wish to force specialization on them. To do so, one needs to have those type annotations. In fact, we may wish to force aggressive constant propagation, so all the character handling gets compiled away.

Comment (araujoms) [reply stdlib/LinearAlgebra/src/symmetric.jl]: They are good suggestions, I added them. It turned out that I can't reuse the kernel for triangular matrices, because since I'm unwrapping them I can no longer rely on the zero elements being zero, so I have to do that manually. I've added a dedicated kernel for them, which is anyway slightly faster. I just hope there's someone actually interested in taking Kronecker products of triangular matrices.

I did deviate from your suggestion in one point: I added not only a `kron` method, but also a `kron!` method. Since `LinearAlgebra` currently exports them for the generic case, it would be weird if they are suddenly missing for some arguments.

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/symmetric.jl]: Hm, the triangular ones are pretty peculiar. I was roughly thinking one could reuse some code, but forgot about the triangle to-be-ignored... Feels a bit like code bloat, TBH. I'll let others chime in on whether they believe the triangular stuff is worth it. Personally, I'd leave it with the symmetric/hermitian optimization.

Comment (araujoms) [reply stdlib/LinearAlgebra/src/symmetric.jl]: It doesn't seem like others will chime in.

Comment (jishnub) [reply stdlib/LinearAlgebra/src/symmetric.jl]: Let's just do the hermitian for now

Comment (araujoms) [reply stdlib/LinearAlgebra/src/symmetric.jl]: :facepalm: I wasted a lot of time doing the triangular case only because @dkarrasch asked. I have no use for it. Next time please only ask for features you actually want.

Comment (araujoms) [reply stdlib/LinearAlgebra/src/symmetric.jl]: Well?

**Comment by ViralBShah** [comment]
@araujoms Is this PR ready from your perspective?

**Comment by araujoms** [comment]
Yes.

**Comment by ViralBShah** [comment]
I'll wait for @jishnub to chime in and hope to get this merged next week. Thanks!

**Comment by araujoms** [comment]
Oh, ok, thanks a lot.

File: stdlib/LinearAlgebra/src/triangular.jl
Comment (jishnub) [inline stdlib/LinearAlgebra/src/triangular.jl:822]:
```suggestion
                    C[inB+l, jnB+k] = _zero(C, inB+l, jnB+k)
```
and similarly, elsewhere. Currently, on this PR, the following is broken:
```julia
julia> U2 = UpperTriangular(reshape(fill(ones(2,2),4), 2, 2))
2×2 UpperTriangular{Matrix{Float64}, Matrix{Matrix{Float64}}}:
 [1.0 1.0; 1.0 1.0]  [1.0 1.0; 1.0 1.0]
         ⋅           [1.0 1.0; 1.0 1.0]

julia> kron(U2, U2)
ERROR: MethodError: no method matching zero(::Type{Matrix{Float64}})
The function `zero` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  zero(::Type{Union{}}, Any...)
   @ Base number.jl:310
  zero(::Type{Missing})
   @ Base missing.jl:104
  zero(::Type{Dates.DateTime})
   @ Dates ~/Downloads/julia-674fcf7a91-linux-x86_64/julia-674fcf7a91/share/julia/stdlib/v1.11/Dates/src/types.jl:458
  ...

Stacktrace:
 [1] _triukron!(C::Matrix{Matrix{Float64}}, A::Matrix{Matrix{Float64}}, B::Matrix{Matrix{Float64}})
   @ LinearAlgebra ~/Downloads/julia-674fcf7a91-linux-x86_64/julia-674fcf7a91/share/julia/stdlib/v1.11/LinearAlgebra/src/triangular.jl:790
 [2] kron!
   @ ~/Downloads/julia-674fcf7a91-linux-x86_64/julia-674fcf7a91/share/julia/stdlib/v1.11/LinearAlgebra/src/triangular.jl:767 [inlined]
 [3] kron(A::UpperTriangular{Matrix{Float64}, Matrix{Matrix{Float64}}}, B::UpperTriangular{Matrix{Float64}, Matrix{Matrix{Float64}}})
   @ LinearAlgebra ~/Downloads/julia-674fcf7a91-linux-x86_64/julia-674fcf7a91/share/julia/stdlib/v1.11/LinearAlgebra/src/triangular.jl:757
 [4] top-level scope
   @ REPL[51]:1
```
I think this should be fixed by this change.

**Review by jishnub — CHANGES_REQUESTED** [review]
Looks ok to me, but we would want to add tests for matrix-valued matrices (which is broken on this PR currently), and also for matrices where the parent isn't initialized corresponding to the structural zeros (which works, but doesn't seem to be tested for).
E.g.:
```julia
julia> M = Matrix{BigInt}(undef,2,2);

julia> M[1,1] = M[2,2] = M[1,2] = 2;

julia> U = UpperTriangular(M);

julia> kron(U, U)
4×4 Matrix{BigInt}:
 4  4  4  4
 0  4  0  4
 0  0  4  4
 0  0  0  4
```
Some such tests would prevent future breakages.

File: stdlib/LinearAlgebra/src/triangular.jl
Comment (araujoms) [reply stdlib/LinearAlgebra/src/triangular.jl:822]: Ops, indeed, thanks for catching that. However, I don't think this is a full solution; if we do a `Matrix{Matrix{Matrix}}`, that is,
```julia
c=reshape(fill(ones(2,2),4), 2, 2)
d = [c for i=1:2, j=1:2]
```
then we still have a crash, and currently `kron` can handle this.

Comment (araujoms) [reply stdlib/LinearAlgebra/src/triangular.jl:822]: Upon further investigation, I saw that there is also a problem with `Symmetric`:
```julia
julia> s=Symmetric(reshape(fill(ones(2,2),4), 2, 2))
2×2 Symmetric{AbstractMatrix, Matrix{Matrix{Float64}}}:
 [1.0 1.0; 1.0 1.0]  [1.0 1.0; 1.0 1.0]
 [1.0 1.0; 1.0 1.0]  [1.0 1.0; 1.0 1.0]

julia> kron(s,s)
ERROR: MethodError: no method matching symmetric_type(::Type{Any})
The function `symmetric_type` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  symmetric_type(::Type{T}) where T<:Number
   @ LinearAlgebra ~/local/julia/usr/share/julia/stdlib/v1.11/LinearAlgebra/src/symmetric.jl:95
  symmetric_type(::Type{T}) where {S<:(AbstractMatrix), T<:AbstractMatrix{S}}
   @ LinearAlgebra ~/local/julia/usr/share/julia/stdlib/v1.11/LinearAlgebra/src/symmetric.jl:92
  symmetric_type(::Type{T}) where {S<:Number, T<:AbstractMatrix{S}}
   @ LinearAlgebra ~/local/julia/usr/share/julia/stdlib/v1.11/LinearAlgebra/src/symmetric.jl:89
  ...

Stacktrace:
 [1] symmetric_type(::Type{Matrix{Any}})
   @ LinearAlgebra ~/local/julia/usr/share/julia/stdlib/v1.11/LinearAlgebra/src/symmetric.jl:87
 [2] Symmetric(A::Matrix{Any}, uplo::Symbol)
   @ LinearAlgebra ~/local/julia/usr/share/julia/stdlib/v1.11/LinearAlgebra/src/symmetric.jl:62
 [3] kron(A::Symmetric{AbstractMatrix, Matrix{Matrix{Float64}}}, B::Symmetric{AbstractMatrix, Matrix{Matrix{Float64}}})
   @ LinearAlgebra ~/local/julia/usr/share/julia/stdlib/v1.11/LinearAlgebra/src/symmetric.jl:532
 [4] top-level scope
   @ REPL[105]:1

```
Note that the `eltype(s)` is `AbstractMatrix`. When this gets fed into `promote_op` it goes crazy.

Honestly I think it's better to just change the methods to
```julia
function kron(A::Symmetric{T}, B::Symmetric{S}) where {T<:Number,S<:Number}
```
and so on, so that we remove these recursive matrices from consideration.

**Comment by araujoms** [comment]
I've added the tests you asked and removed the recursive matrices from consideration.

**Comment by araujoms** [comment]
I've realized that the recursive case can be handled by `zero` thanks to #51458; this makes supporting the triangular types easy. However, `Symmetric` would still fail because of the bug with the `eltype` of `Symmetric(reshape(fill(ones(2,2),4), 2, 2))`. I'm not sure if fixing that is worth it.

**Comment by araujoms** [comment]
Gentle bump.

**Comment by araujoms** [comment]
Well?

**Comment by araujoms** [comment]
Please?

**Comment by ViralBShah** [comment]
@jishnub The merging is blocked. I believe your review is holding it back (and perhaps it just needed to be marked as resolved). Can this be merged?

**Review by jishnub — APPROVED** [review]
Sorry for the delay, looks good to me

**Pull request merged**
